### ISSUE TITLE
Timestamp timezone

### ISSUE THREAD

**Issue by jrstrunk** [gleam-lang/time#4]
Hey Louis, just saw you merged #1 but haven't published yet. I read over some of the code again and noticed that at the beginning of the Timestamp module it says "... represented as an amount of time since 00:00:00 UTC on 1 January 1970...". The `to_rfc3339` function seems to conflict with this since it takes an offset.

We might want to settle on whether the Timestamp type always represents UTC time or not and either update the docs or the `to_rfc3339` function.

**Comment by lpil** [comment]
The timestamp is always UTC.

RFC3339 timestamps have multiple ways of representing the same point in time as it includes an offset part. Depending on what offset is desired a different set of numbers is used for the same point in time.

The function takes a timestamp and an offset because Unix timestamps are always UTC, so you need to also state what offset you wish to use.

**Comment by jrstrunk** [comment]
Oh, I see. That makes sense; it is all consistent. It is just the API's semantics that seemed a little strange to me. The `to_rfc3339` function is a two-in-one function that changes the time value _and_ generates a string representation of it.

Given that more expressive types will eventually be added, having a more atomic API where time conversions and string generations are done separately may be clearer. This would make for fewer magic `0` values in a codebase where users want to display a Timestamp as a string without converting it to another offset. It would also mean that for now, the Timestamp type only supports a subset of RFC3339 where the offset is always `Z`, but later on, a user could do something like `timestamp_var |> timestamp.to_offset(60) |> datetime.to_rfc3339` where the `timestamp.to_offset` function converts it to a more expressive type which supports all of RFC3339. 

What do you think?

**Comment by lpil** [comment]
It is not changing the time, it continues to represent the same time and the offset argument specifies which RFC3339 format is used to represent that time.

Having folks change the timestamp and then format it would result invalid timestamps being generated while also not removing the need for the second argument.

It is not a magic number, it's an offset. An RFC3339 timestamp (and also the similar ISO timestamp) represent a time _and_ an offset, so both are always required.

**Comment by jrstrunk** [comment]
In my quick example, the intention was for `timestamp_var |> timestamp.to_offset(60)` to return a DateTime type so an invalid Timestamp would not be generated. This function could not be implemented yet, just thinking long term.

> It is not a magic number, it's an offset. An RFC3339 timestamp (and also the similar ISO timestamp) represents a time and an offset, so both are always required.

The Timestamp type already has an assumed offset that can be used, the user does not necessarily need to provide one.

**Comment by lpil** [comment]
Not providing one is still picking one, an implicit choice of zero. There is no avoiding that these timestamp formats always include an offset, so only options are between permitting the programmer to use any offset, or to force the programmer to use one specific one.

This is to be a foundational library, so I think it would be good to support projects that need to be able to specify offsets.

If you believe that no Gleam project needs to be able to render timestamps with offsets other than 0 you will need to provide a compelling argument for this.

**Comment by jrstrunk** [comment]
I am not suggesting to remove any functionality, just separate it.

I should try to be clearer about the bottom line. The `to_rfc3339` function does two things at once, but it may be a better experience for the consumers of this library to have those two things separated into two different, composable functions: one for assigning an offset to a Timestamp (converting it to a DateTime type in the process) and another for converting it to a string.

> If you believe that no Gleam project needs to be able to render timestamps with offsets other than 0

If a DateTime type will be added later, then I don't think it makes sense to render timestamps with offsets other than 0. Are you not planning to add a DateTime type later?

> No, it does not. Offsets and zones have no relevance to Unix time.

> ... represented as an amount of time since 00:00:00 UTC on 1 January 1970...

How can both of these things be true? It is stated that UNIX represents UTC time (but skipping leap seconds).

**Comment by jrstrunk** [comment]
I see what you are saying about the offset now. I always thought of timestamps as having an implicit 0 offset, but I see how that is technically wrong. Thinking about the `to_rfc3339` function's design this way makes much more sense. It was just a little strange to me. Thanks for walking through it, sorry there was so much back and forth. Ultimately I just want to see this package succeed.

> What would you say makes the split API a better experience for the user?

I just think that composable APIs are generally easier to work with and more readable. Would you agree?

Now that it is established that Timestamps and DateTimes are distinctly different and not directly comparable, can we agree that the `to_rfc3339` function does two things: it 1. converts the timestamp into a datetime, then 2. renders that datetime as a string?

**Comment by lpil** [comment]
> I just think that composable APIs are generally easier to work with and more readable. Would you agree?

No, I don't find it to be so. I find it roughly the same, but the verbosity and needing to know about multiple functions makes it slightly harder to understand.

> Now that it is established that Timestamps and DateTimes are distinctly different and not directly comparable, can we agree that the to_rfc3339 function does two things: it 1. converts the timestamp into a datetime, then 2. renders that datetime as a string?

I don't think it does that, no. The API of the function is Timestamp to RFC3339, there's no date-time structure there.

**Comment by jrstrunk** [comment]
> No, I don't find it to be so. I find it roughly the same, but the verbosity and needing to know about multiple functions makes it slightly harder to understand.

Good point, discretion is needed here based on the API's goals. Composable APIs are more powerful but also more difficult. My statement was too broad, we should consider composability concretely within the context of this package.

> I don't think it does that, no.

Hmm am I missing something? Is it not true that a UNIX timestamp is a duration value while RFC3339 is a format for datetime values? Knowing the epoch of a duration allows you to convert it to a datetime value, but a duration cannot be directly formatted in the RFC3339 format, right?

I am having trouble finding another language besides Elm with an API to convert a timestamp value directly to a formatted datetime string. I cannot help but think this is because Elm has very minimal time types, and the `Posix` type is at the forefront. This leads me to a foundational question:

When Gleam's standard time library is fully mature, will it support a variety of different time types (e.g. Date, Time, DateTime, Timestamp, etc., separately), or will it only provide a minimal number of types like Elm?

**Comment by lpil** [comment]
> Hmm am I missing something? Is it not true that a UNIX timestamp is a duration value while RFC3339 is a format for datetime values? Knowing the epoch of a duration allows you to convert it to a datetime value, but a duration cannot be directly formatted in the RFC3339 format, right?

Sorry, I don't know what you mean. A `Timestamp` and a RFC3339 string are both the same thing- an exact point in time. Neither are durations. The way they differ is the they use, and some properties that come from that (such as RFC3339 having multiple ways to represent the same point in time, while `Timestamp` has a canonical representation).

> When Gleam's standard time library is fully mature, will it support a variety of different time types (e.g. Date, Time, DateTime, Timestamp, etc., separately), or will it only provide a minimal number of types like Elm?

The goal is to make people always use timestamps as it is the most efficient data structure with the smallest number of sharp edges for manipulating time.

Calendar data types will be provided but their use for anything other than displaying to humans will be discouraged as much as possible. They are complex and very easy to misuse, which is why time is typically difficult in programming.

**Comment by jrstrunk** [comment]
Okay, I see your reasoning. That resolves my reservations, thank you!
